nextcloud/spreed
# Fix permissions not honoured if updated right before joining a call

When the local participant joins a call the sent media is based on the last fetched permissions. However, once the call is joined the sent media is based on the participant permissions updates notified by the signaling messages.

When the HPB is used the [`participantPermissions` property](https://github.com/nextcloud/spreed/blob/43ebbd09a2b9139601b945a1ab77fc95a30a5290/src/utils/webrtc/webrtc.js#L981) of the [`participants->update` signaling messages](https://github.com/nextcloud/spreed/blob/79455819db19ae1c01ca5b471441e345a23a53ed/src/utils/signaling.js#L1268) is used. Those messages are sent when the permissions are modified and also when a participant joins a call, but in this last case it did not include the expected property.

Due to this, if the permissions of the local participant were updated right before the participant joined the call and the participant data was not fetched again yet the previous permissions would be used. If the message sent when the permissions were updated was received before actually joining the call then it would be ignored, and although the message sent due to joining the call would be processed the WebUI would not notice the permission change, as the message did not provide them.

To solve that now the `participants->update` message sent when joining a call also includes the permissions. This way, even if the previous permissions are initially used when joining the call as soon as the signaling message sent due to joining the call is received the WebUI will adjust the sent media to the new permissions.

The tests below use guests because updating the permissions does not cause the participant list to be fetched again* (because they can not fetch it), so there is more time to trigger the race condition. Nevertheless, it could happen as well with regular users, but it is easier to reproduce it with guests.

*Probably the conversation data should be fetched again when guests are used and the current participant changes, but that could be done in another pull request.

## How to test (scenario 1)

- Setup the HPB
- Open a conversation as a moderator
- Grant audio and video permissions by default
- In a private window, join the conversation as a guest
- Start a call with audio and video, but do not join the call, wait in the device checker
- In the original window, remove audio and video permissions for the guest
- In the private window, join the call
- In the original window, join the call without audio nor video

### Result with this pull request

In the private window, eventually (after a forced reconnection) the UI shows the media as disabled, `OCA.Talk.SimpleWebRTC.webrtc.localStreams` is empty and the local participant does not try to send media to the HPB
- Sometimes the guest does not automatically join the call again after the forced reconnection; that still needs to be investigated, but it is most likely unrelated and a different bug

### Result without this pull request

In the private window, the UI shows the media as enabled (even if eventually the tooltip says _You are not allowed to enable the microphone/camera_), `OCA.Talk.SimpleWebRTC.webrtc.localStreams` is not empty and the local participant is trying to send media to the HPB (there is a loading spinner on the local participant, and the HPB logs show messages like _XXX is not allowed to offer video, ignoring (permission "publish-audio" not found)_)



## How to test (scenario 2)

- Setup the HPB
- Open a conversation as a moderator
- In a private window, join the conversation as a guest
- In the original window, remove audio and video permissions for the guest
- Wait more than 30 seconds to ensure that the guest received the permissions update
- In the private window, start a call with audio and video, but do not join the call, wait in the device checker
- In the original window, grant audio and video permissions for the participant
- In the private window, join the call
- In the original window, join the call without audio nor video

### Result with this pull request

In the private window, eventually (after a forced reconnection) it is possible to enable the media and receive it in the original window

### Result without this pull request

In the private window, the UI shows the media as disabled and it is not possible to enable it (_You are not allowed to enable the microphone/camera_ is eventually replaced by _No microphone/camera_, even if a microphone or a camera is selected)
